src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Refactoring/CodeActionEndpointTest.cs

```diff
@@ -0,0 +1,230 @@
+using System.Linq;
+using System.Threading;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Test.Refactoring
+{
+    public class CodeActionEndpointTest : LanguageServerTestBase
+    {
+        private readonly DocumentResolver EmptyDocumentResolver = Mock.Of<DocumentResolver>();
+
+        [Fact]
+        public async Task Handle_NoDocument()
+        {
+            // Arrange
+            var documentPath = "C:/path/to/Page.razor";
+            var codeActionEndpoint = new CodeActionEndpoint(new RazorCodeActionProvider[] { }, Dispatcher, EmptyDocumentResolver, LoggerFactory);
+            var request = new CodeActionParams()
+            {
+                TextDocument = new TextDocumentIdentifier(new Uri(documentPath)),
+                Range = new Range(new Position(0, 1), new Position(0, 1)),
+            };
+
+            // Act
+            var commandOrCodeActionContainer = await codeActionEndpoint.Handle(request, default);
+
+            // Assert
+            Assert.Null(commandOrCodeActionContainer);
+        }
+
+        [Fact]
+        public async Task Handle_UnsupportedDocument()
+        {
+            // Arrange
+            var documentPath = "C:/path/to/Page.razor";
+            var codeDocument = CreateCodeDocument("@code {}");
+            var documentResolver = CreateDocumentResolver(documentPath, codeDocument);
+            codeDocument.SetUnsupported();
+            var codeActionEndpoint = new CodeActionEndpoint(new RazorCodeActionProvider[] { }, Dispatcher, documentResolver, LoggerFactory);
+            var request = new CodeActionParams()
+            {
+                TextDocument = new TextDocumentIdentifier(new Uri(documentPath)),
+                Range = new Range(new Position(0, 1), new Position(0, 1)),
+            };
+
+            // Act
+            var commandOrCodeActionContainer = await codeActionEndpoint.Handle(request, default);
+
+            // Assert
+            Assert.Null(commandOrCodeActionContainer);
+        }
+
+        [Fact]
+        public async Task Handle_NoProviders()
+        {
+            // Arrange
+            var documentPath = "C:/path/to/Page.razor";
+            var codeDocument = CreateCodeDocument("@code {}");
+            var documentResolver = CreateDocumentResolver(documentPath, codeDocument);
+            var codeActionEndpoint = new CodeActionEndpoint(new RazorCodeActionProvider[] { }, Dispatcher, documentResolver, LoggerFactory);
+            var request = new CodeActionParams()
+            {
+                TextDocument = new TextDocumentIdentifier(new Uri(documentPath)),
+                Range = new Range(new Position(0, 1), new Position(0, 1)),
+            };
+
+            // Act
+            var commandOrCodeActionContainer = await codeActionEndpoint.Handle(request, default);
+
+            // Assert
+            Assert.Empty(commandOrCodeActionContainer);
```

Comment:
What's the expected behavior difference between an empty container and `commandOrCodeActionContainer` being null?

Reply:
OmniSharp doesn't differentiate between null and an empty array. I still use this to check dispatch success, however; returning null means it never reached dispatch and returning an empty array means that none of the providers deemed it an applicable scenario.